# Numpy

*NumPy (Numerical Python) is an open-source Python library essential for scientific computing. It provides a powerful n-dimensional array object, tools for integrating C/C++ and Fortran code, and useful linear algebra, Fourier transform, and random number capabilities. At its core, NumPy offers a high-performance multidimensional array object and functions for manipulating these arrays efficiently, making it a cornerstone for data analysis and numerical computation in Python.*

### Fucntions:
- numpy.mean
- numpy.std 
- numpy.dot
- numpy.reshape
- numpy.where

### numpy.mean:
*The mean function computes the arithmetic mean along the specified axis.*

In [1]:
import numpy as np
gene_expression = np.array([10, 12, 8, 15, 9])
mean_expression = np.mean(gene_expression)
mean_expression

10.8

- Explanation:
*This calculates the average of the gene expression levels, resulting in 10.8.*





In [2]:
data = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
mean_rows = np.mean(data, axis=1)
mean_rows

array([2., 5., 8.])

- Explanation: 
*This computes the mean for each row, resulting in an array [2.0, 5.0, 8.0].*

In [3]:
data = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
mean_columns = np.mean(data, axis=0)
mean_columns

array([4., 5., 6.])

- Explanation: 
*This computes the mean for each column, resulting in an array [4.0, 5.0, 6.0].*

### numpy.std:
*The std function computes the standard deviation along the specified axis.*

In [4]:
gene_expression = np.array([10, 12, 8, 15, 9])
std_expression = np.std(gene_expression)
std_expression

2.481934729198171

- Explanation:
*This calculates the standard deviation of the gene expression levels, which is a measure of variability or dispersion.*

In [5]:
data = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
std_rows = np.std(data, axis=1)
std_rows

array([0.81649658, 0.81649658, 0.81649658])

- Explanation: 
*This computes the standard deviation for each row.*


In [6]:
data = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
std_columns = np.std(data, axis=0)
std_columns

array([2.44948974, 2.44948974, 2.44948974])

- Explanation: 
*This computes the standard deviation for each column.*

### numpy.dot:
*The dot function computes the dot product of two arrays. For 2-D vectors, it is equivalent to matrix multiplication.*

In [ ]:
vector1 = np.array([1, 2, 3])
vector2 = np.array([4, 5, 6])
dot_product = np.dot(vector1, vector2)
dot_product

- Explanation: 
*This calculates the dot product (14 + 25 + 3*6), resulting in 32.*

In [ ]:
matrix1 = np.array([[1, 2], [3, 4]])
matrix2 = np.array([[5, 6], [7, 8]])
result = np.dot(matrix1, matrix2)
result

- Explanation: 
*This performs matrix multiplication, resulting in a new matrix.*

In [ ]:
profile1 = np.array([0.1, 0.2, 0.3])
profile2 = np.array([0.4, 0.5, 0.6])
similarity = np.dot(profile1, profile2)
similarity

- Explanation: 
*This calculates the similarity between two gene expression profiles, resulting in 0.32.*

### numpy.reshape:
*The reshape function is used to give a new shape to an array without changing its data.*

In [7]:
array = np.array([1, 2, 3, 4, 5, 6])
reshaped_array = np.reshape(array, (2, 3))
reshaped_array

array([[1, 2, 3],
       [4, 5, 6]])

- Explanation: 
*This reshapes the 1D array into a 2x3 2D array.*

In [ ]:
array = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
reshaped_array = np.reshape(array, (2, 2, 2))
reshaped_array

- Explanation: 
*This reshapes the 2D array into a 2x2x2 3D array*



In [8]:
array = np.array([[1, 2, 3], [4, 5, 6]])
flattened_array = np.reshape(array, -1)
flattened_array

array([1, 2, 3, 4, 5, 6])

- Explantion:
*This flattens the 2D array into a 1D array.*

### numpy.where:
*The where function returns elements chosen from x or y depending on condition.*

In [9]:
array = np.array([1, 2, 3, 4, 5])
indices = np.where(array > 3)
indices

(array([3, 4], dtype=int64),)

- Explanation: 
*This returns the indices of elements in the array that are greater than 3.*

In [10]:
array = np.array([1, 2, 3, 4, 5])
modified_array = np.where(array > 3, 10, array)
modified_array

array([ 1,  2,  3, 10, 10])

- Explanation: 
*This replaces elements greater than 3 with 10.*

In [11]:
array = np.array([1, 2, 3, 4, 5, 2, 3])
indices = np.where(array == 2)
indices

(array([1, 5], dtype=int64),)

- Explanation: 
*This returns the indices of elements in the array that are equal to 2.*